### ISSUE TITLE
Create a `format` Registry

### ISSUE THREAD

**Issue by webron** [OAI/OpenAPI-Specification#845]
In reference to #607 and #811.

Instead of adding additional formats to the spec, we're considering two options - either a set of formats in a supporting guidelines document or a formal OAI format registry with a set of guidelines how to enter new formats to it. The registry will serve as an official repository for these formats and tools could use it as a reference. The @OAI/tdc is currently leaning towards the option of the repository.

This ticket is a reminder to tackle it and finalize the approach and the guidelines for either case.

**Comment by MikeRalphson** [comment]
It may be worth recommending that format strings only contain lower-case letters, digits and hyphens, or some other restriction, to draw attention to the fact it is not intended as an [essay on the content of the property](https://github.com/Mermade/openapi-specification-extensions/blob/master/formats/github.tsv#L48) or an enum, pattern or template.

**Comment by handrews** [comment]
Relevant: https://github.com/json-schema-org/json-schema-spec/issues/563

**Comment by MikeRalphson** [comment]
Proposed candidates discussed in other issues (work-in-progress):

|Type(s)|Format|Description|Issue|
|---|---|---|---|
string|int64s|64-bit integer held as a string for interoperability reasons|#1517
integer|int53|53-bit integer|#1517
integer|uint8|Unsigned 8-bit integer (*To help with misunderstandings of the current `byte` format*)
number\|string|decimal|Fixed-point decimal numbers|#889
string|uuid|UUID v4 format|#542

**Comment by MikeRalphson** [comment]
@handrews - I like the idea of potentially delegating responsibility for formats to a JSON Schema-led registry and collaborating on that. 

> For now, we'll keep a small standard set in one of the major specification drafts.

How close would we be on the "core" OAS [formats](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#data-types) ?

**Comment by handrews** [comment]
@MikeRalphson regarding "core" OAS formats:

* "byte" should be done with `"contentEncoding": "base64"` as of draft-07
* "binary" should be done with `"contentMediaType": "application/octet-stream" as of draft-07
* Both of the above have existed in some form for much longer than draft-07 under various names
* The numeric formats are reasonable (to me) but we've never really sorted that out
* "password" should be done through a UI vocabulary, so probably not a `format`

@darrelmiller regarding #607:
* "decimal" is proposed for JSON Schema _but as a string format_ if someone who understands the use cases and concerns will just write  $*%$@ PR for it- it could have been in draft-07 if anyone who was asking about it would step up.  I don't think it's possible to reliably implement it as a number format so I'm confused there.
* "uuid" Is this just a UUID or a UUID URN?  The latter should be `{"type": "string", "format": "uri", "pattern": "^urn:uuid:"}`
* "duration" is one that I'd like to have, we already have the other date and time ones
* "base64url" should also be a `"contentEncoding": "base64url"`, not `format`

**Comment by cyberphone** [comment]
What's the difference between "byte" and "binary"?
In most newer standards, byte-arrays are encoded as Base64Url rather than Base64.

`BigInteger` support is currently available in Java, .NET, and probably in a bunch of other platforms as well.  Java and .NET use entirely different `BigInteger` serialization schemes.

I would separate a possible `BigNumber` type from `Decimal`/`Money` because the latter is based on decimal arithmetic and usually do not come with exponents.  If an application needs exponents, `BigNumber` would be the logical choice.

**Comment by cyberphone** [comment]
The JSON-P API for Java use the following algorithm for long and a similar one for other extended numeric types:
```java
if (value >= 9007199254740992 || value <= -9007199254740992)
    serializeAsString(value);
else
    serializeAsNumber(value);
```
IMO, it is bad idea but they claim it is the "industry standard" as well as the correct interpretation of the JSON RFC.

**Comment by ioggstream** [comment]
I propose the fixed-point numeric `decimal` as per

  -  "DECIMAL"  SQL standard "ISO/IEC 9075-2: 2016 12 15"

**Comment by cyberphone** [comment]
> I propose the fixed-point numeric decimal as per
>  "DECIMAL" SQL standard "ISO/IEC 9075-2: 2016 12 15"

Don't you just love standards organizations that want money for their work? https://webstore.iec.ch/publication/59685

**Comment by ioggstream** [comment]
"Use the (draft)[http://jtc1sc32.org/doc/N2301-2350/32N2311T-text_for_ballot-CD_9075-2.pdf], Luke"  ;)

**Comment by grokify** [comment]
Specifically for date/times, given the large variety of formats that exist in the wild, would it be more flexible and easy to implement in code generation tools to support a format specification than a registry enumeration? For example, a format could use `strftime` or [Go's time format](https://golang.org/pkg/time/#pkg-constants).

**Comment by ioggstream** [comment]
>Specifically for date/times, given the large variety of formats
> that exist in the wild, ... could use strftime or Go's time format.
@grokify iiuc date-time uses JSON Schema's date-time which is based on RFC 3339 which is a subset of ISO 8601.

**Comment by grokify** [comment]
> > Specifically for date/times, given the large variety of formats that exist in the wild, ... could use strftime or Go's time format.
>
> @grokify iiuc date-time uses JSON Schema's date-time which is based on RFC 3339 which is a subset of ISO 8601.

Restricting date-time to RFC 3339 is useful when creating an API, but is not as flexible for existing APIs that may not use it.

When looking at the registry provided by @MikeRalphson, there are a number of non-comforming date-times:

https://github.com/Mermade/openapi-specification-extensions/blob/master/formats/combined.tsv

A recent example I'm attempting to use is [Insightly's API](https://api.insight.ly/v2.2/Help#!/Overview/Technical_Details). They provide an [OpenAPI 2.0 spec](https://api.insight.ly/v2.2/swagger/docs/v2.2) that specifies `date-time` but they do not use RFC 3339. Their formats are listed in their documentation:

* Query string: `M/d/yyyy h:mm:ss AM/PM` - `11/7/2015 8:07:05 AM`
* Object data: `yyyy-MM-dd HH:mm:ss` - `2015-04-10 21:15:00`

This causes the Swagger Codgen SDK I'm attempting to use to raise exceptions with response data. In order to continue to use Swagger Codgen right now, my current thinking is one of the following:

1. Modify Swagger Codegen to use a simple `string` type
2. Modify Swagger Spec and Swagger Codegen to use JSON Schema x-date-time-format property with custom parsing

Aside from formal spec support, I'm leanding to #1 for short term solution and #2 for longer-term solution.

What would your recommendation be for Swagger Codegen here?

**Comment by MikeRalphson** [comment]
> When looking at the registry provided by @MikeRalphson, there are a number of non-comforming date times:

https://github.com/Mermade/openapi-specification-extensions/blob/master/formats/combined.tsv

> With the OAI format registry approach, would each of the date / time format's be listed as a separate format in the registry, e.g. not date-time?

Just for clarity, the list linked to above is not a "registry", it's merely a survey (and and out of date one at that) of `format` values used in the wild. There is no expectation that values from that list would be added unreviewed to the proposed registry. There is no guarantee that a `format` value used by one provider has the same meaning across all APIs which use it. `format` will remain an open-ended property where you can define your own values.

With regard to the Insightly API, both of your approaches sound fine, though I would add:

0. Notify upstream API provider of the error in their OAS document

Certainly the meaning of `date-time` inherited from JSON schema is very unlikely to change, and unless there is a common standard for custom date formatting (as opposed to Go's, Javascript's, C's `strftime` etc) then having such a format (or additional property) would not meet OAS's goal of being language-neutral.

**Comment by handrews** [comment]
@MikeRalphson @grokify it is correct that `date-time` and the other RFC 3339-based `format` values will not change.

As I noted in @grokify's issue filed on JSON Schema, our solution to the problem of managing `format` extensibility is being tracked at https://github.com/json-schema-org/json-schema-spec/issues/563

While I think it's likely that we will add some duration formats alongside the existing date and time formats, that would be the extent of any additions to the JSON Schema spec itself.  The myriad other possible ways to express these concepts would be handled as extensions.

**Comment by grokify** [comment]
@MikeRalphson 

> Notify upstream API provider of the error in their OAS document

I'm reluctant to do this because the current solution of falling back to a `string` is a lossy and that information is important. I prefer to have this information and then find some way to handle it myself. For Option 1, it's simple to write a simple script that removes that parameter but then I also know which properties should have additional processing for Option 2. Without this information, Option 2 would be more difficult to know which properties to apply it on a systemic basis.

**Comment by MikeRalphson** [comment]
@grokify I would expect the API provider to replace, not remove the misleading and incorrect `format` value with one (or more!) of their own.

**Comment by grokify** [comment]
> @grokify I would expect the API provider to replace, not remove the misleading and incorrect format value with one (or more!) of their own.

@MikeRalphson To me, it seems there is no standard approach to replace this info at this time. How would you recommend this spec be updated given the description of the API?

My understanding is that Swagger 2.0 Spec, and possibly OpenAPI 3.0 Spec, was not designed to specify all possible APIs. If this is the case, some APIs may not be adequately specified by the spec. With Swagger 2.0 Spec, some features in an API I manage could not be fully specified so we had to use a non-valid spec in our Swagger UI implementation.

**Comment by MikeRalphson** [comment]
> To me, it seems there is no standard approach to replace this info at this time. How would you recommend this spec be updated given the description of the API?

@grokify With any non-`date-time` value for `format` (which is an extensible property, anyone can define their own values, e.g `insightly-date-query` / `insightly-date-object`). `date-time` means what it means, and shouldn't be used where it is inappropriate. I would expect codegen tools to have some kind of mechanism for overriding the behaviour of unknown `format` values, but this might not be the case.

> My understanding is that Swagger 2.0 Spec, and possibly OpenAPI 3.0 Spec, was not designed to specify all possible APIs.

Correct.

> With Swagger 2.0 Spec, some features in an API I manage could not be fully specified so we had to use a non-valid spec in our Swagger UI implementation.

Did you have to hack Swagger-UI? I would recommend using the supported specification extensions to do this, and to not deviate from the spec itself. I don't see how having an invalid OAS document helps you at all.

**Comment by grokify** [comment]
@MikeRalphson Ah, I see `format` is an open value now. I would probably recommend using strftime, Go format, or some other format rather than something like `insightly-date-query` which would require an external reference. Something like the following may be more useful in my mind.

```
"format": "strftime"
"x-format": "%Y-%m-%d %H:%M:%S"
```
> Did you have to hack Swagger-UI? I would recommend using the supported specification extensions to do this, and to not deviate from the spec itself. I don't see how having an invalid OAS document helps you at all.

I've hacked both Swagger UI and json-editor. If you read the referenced GitHub issue, there was no supported extension in the 2.0 time frame.

As for how an invalid OAS spec helps, the goal of Swagger UI is to support UI based access to APIs. If not all APIs are supported by OAS spec and hacking Swagger UI can make those APIs available in the UI, it is beneficial to support that use case. We have a different version of our spec that validates that we use for code generation. Different specs for different use cases.

**Comment by MikeRalphson** [comment]
@grokify glad we're on the same page now. Yes extending the `schema object` with something like `x-format-specifier` would be a suitable approach.

> If you read the referenced GitHub issue, there was no supported extension in 2.0.

Which issue? If you mean in the spec, specification/vendor extensions **are** supported in OAS 2.0 in the same way as in 3.0.x

> As for how an invalid OAS spec helps...

This isn't an argument for an **invalid** OAS document, it's an argument for an **extended** one.

**Comment by ioggstream** [comment]
@grokify imho 
> "format": "strftime"
> "x-format": "%Y-%m-%d %H:%M:%S"
two wrongs won't make it right ;)

**Comment by grokify** [comment]
> > As for how an invalid OAS spec helps...
> 
> This isn't an argument for an invalid OAS document, it's an argument for an extended one.

An invalid spec is not a perfect solution and there is no general desire to have an invalid spec. It is a compromise given the state of the ecosystem. Sometimes having a valid spec for a not-well supported capability means more invasive changes and maintenance are needed in tooling and ecosystem tools. Sometimes, these changes can be justified, but other times it's harder to do so. In the specific scenario here, having two specs, a valid one for interoperability and another specifically to be used in one app without direct consumption seemed like a reasonable compromise.

> > "format": "strftime"
> > "x-format": "%Y-%m-%d %H:%M:%S"
> 
> two wrongs won't make it right ;)

I'm trying to find a working, practical solution. Another approach I'm thinking of is using `x-pattern` instead of `x-format`. `pattern` is already used for regular expressions which seems more similar to the use case here.

**Comment by grokify** [comment]
> * using alternate format for dates is probably a bad practice
> * that should be blamed and made difficult
> * we should strive to harmonization and not prolification

@ioggstream I understand this, but I'm prioritizing working over perfect. For new APIs, we should certain strive for harmonization and I spend quite a bit of time thinking about this when designing new APIs to be easily spec'd by OpenAPI. However, I also have a requirement to access existing APIs and existing APIs can take some time to update. Using `format` and `x-pattern` are valid and seem like a reasonable compromise.

**Comment by ansonkao** [comment]
Is `time` an officially recognized format, or not?

Above, I see it in the table like so:

Type(s) | Format | Description | Issue
-- | -- | -- | --
string | time | time of day - as defined by partial-time - RFC3339 | #358

But it is not on the official docs next to `date` and `date-time`

**Comment by MikeRalphson** [comment]
@ansonkao 

> Is time an officially recognized format, or not? Above, I see it in the table like so... But it is not on the official docs next to date and date-time

The table above is a list of possible candidates to be added to a formats registry. So `time` is not an "official" format within the OAS, but as `format` is an open-ended property, you may use it today without problems.

**Comment by chulkilee** [comment]
Although openapi spec currently uses the *old* version of JSON Schema.. why should OAS define acceptable formats and maintain registry by itself? Why don't we just use JSON Schema formats, and leave the extensibility to JSON Schema?

JSON schema team is aware of this "format" issue, and seems like the upcoming release changes its behavior a little bit - see https://json-schema.org/work-in-progress/WIP-json-schema-release-notes.html

**Comment by peteraritchie** [comment]
How would a compliant OAS parser deal with additions to this external registry?  I think versioning the spec based on the state of a registry at an instant of time *might not* be the most beneficial.  

I also thought, how would spec authors deal with draft formats...  What I was thinking is that there be a way to specify format/pattern pairs in the components or something.  e.g.

```yaml
  components:
    x-formats:
      html:
        pattern: '<\s*a[^>]*>(.*?)<\s*/\s*a>'
#...
    schemas:
       markup-text:
         type: string
         format: html # a parser would know here to use '<\s*a[^>]*>(.*?)<\s*/\s*a>' to validate a value
```

**Comment by karenetheridge** [comment]
I don't see much gain in distilling formats down to simple regexes. The advantage of the "format" keyword is to be able to specify constraints that are not possible with a simple regexp, such as "date" which not only checks that digits and symbols are in the right place in the string, but also that the value represents a real date (for example "2021-13-13 25:00:00" passes a simple regexp but is not actually a valid date).

FWIW, many json schema implementations do support plugging additional formats into them -- but this cannot be done with data, but with code, due to the nature of format validation -- so there is no standard for doing so, and will be different for every implementation/language/platform.

**Comment by handrews** [comment]
Since OAS 3.1 is out now, with JSON Schema draft 2020-12 support including support for extension vocabularies, **I would recommend closing this and focusing on adding vocabularies to meet the various needs.**  OAI can at some point designate certain vocabularies as preferred without breaking any compatibility guarantees at all.  Or even make certain ones mandatory, which would still be backwards compatible but would require work to ensure that the newly mandatory vocabularies are supported.  Presumably, only well-supported vocabularies would be blessed in that way, though.

For those who do not know the saga of `format` and what happened in the 4+ years since this issue was filed:

* `format` is a giant pain for implementations because the set of formats is open-ended, but was specified in such a way that people expected it to work for validation, although that was a misreading- it was always optional in some ways.
* The exact situation of what was and was not required in terms of `format` support was a huge point of contention fo years.
* `format` was the only place in the spec where a lot of things people wanted fit, so the list of `format` values grew over time, both within and outside of the JSON Schema spec.
* This was extra-messy because there was no registry of `format` values or expected behavior (hence this issue, and I think other people have their own registries somewhere).
* No other keyword behaves remotely like this.  While `contentMediaType` is open-ended, it is explicitly **not** validated and parsing any such embedded document is handled separately from validation, by whoever calls the validator.  No other keyword is even close (other enumerated value keywords, including `contentEncoding`, used a closed set of values).
* Because of all of this, `format` support "in the wild" was _extremely_ variable, ranging from not supported at all, to partial support (the canonical example being validating the "email" format by just checking that there's an `@` somewhere in the middle), to full support.  And each specific format was variable in level of support.

This was a mess, so in draft 2019-09 (which introduced extension vocabularies), we made `format` not validate at all by default.  Technically you can configure things to require it to validate, and then you probably end up in the unreliable situation described above.

The point of vocabulary support is that, as a schema author, you can indicate what extension vocabularies MUST be present in order for your schema to be correctly processed.  This is not possible with `format` values (vocabularies work at the keyword level, not the keyword value level).  The benefit here is that you could (and really should) have multiple vocabularies, each of which replaces a coherent part of `format` with reliable, clear behavior.  For example:

* A date-time vocabulary that supports at least as much as `format` does.  This is easy to validate and I'm sure would be widely supported.
* A newtorking vocabulary (or possibly multiple ones) ip addresses, hostnames, etc.
* Email, with clear guidance on how much syntactic validation is required (because people argue about it endlessly now)
* URI/IRI, perhaps where the keyword takes a list of valid schemes (one of the challenges of URI validation right now is that full syntactical validation is scheme-specific, and the set of schemes is open-ended), and fails validation if the scheme is not recognized, e.g. `"uri": ["http", "https"]` or `"uri": ["data"]` or `"uri": ["urn"]` (I just made this up right now, don't take it too seriously, there are many possible ways to solve this problem!)

I hope this gives a feel for why JSON Schema wants to move away from `format`, and how better options can be added _without having to wait for the JSON Schema spec team to approve anything._

**Comment by peteraritchie** [comment]
> Since OAS 3.1 is out now, with JSON Schema draft 2020-12 support including support for extension vocabularies, **I would recommend closing this and focusing on adding vocabularies to meet the various needs.** OAI can at some point designate certain vocabularies as preferred without breaking any compatibility guarantees at all. Or even make certain ones mandatory, which would still be backwards compatible but would require work to ensure that the newly mandatory vocabularies are supported. Presumably, only well-supported vocabularies would be blessed in that way, though.

...

Agreed.  My comment more of an "in lieu of anything better..." idea.  Support for vocabularies/dialects is a much better way to support something like this.

**Comment by handrews** [comment]
@Bessonov if you have particular questions or concerns that led you to put a "confused" emoji on my comment, I'd be happy to address them.

**Comment by Bessonov** [comment]
Hey @handrews , thanks for pinging! Github emoji :confused: description doesn't make any sense for me. I use it in terms of "unhappy", like ":(". In this particular case I'm unhappy with current state and described challenges. Not with your explanation :heart:

**Comment by MikeRalphson** [comment]
> I'd be happy to update the list [here](https://spec.openapis.org/registry/format/index.html) if someone points me to the source

@baywet, the source of the draft registry is yaml data in the `gh-pages` branch, specifically [here](https://github.com/OAI/OpenAPI-Specification/tree/gh-pages/registries/_format).

**Comment by baywet** [comment]
Thanks for the pointers.

Everyone: I put a PR together at #3167 feel free to give it a look and tell me what I missed :)

**Comment by baywet** [comment]
Update: The main PR has been merged.

There are a couple of follow up PRs you should jump on if you're interested in the specific topics:
- [ ] #3190 
- [ ] #3185 
- [ ] #3186 
- [ ] #3184

**Comment by MikeRalphson** [comment]
I've uploaded a more recent summary of [formats in the wild](https://github.com/Mermade/openapi-specification-extensions/blob/main/formats/2021/formats.tsv) from 2021.

**Comment by lornajane** [comment]
@Speeddymon could you open a new issue to request that please? The issue should be in this repository, and I recommend you include your use cases to help us understand the request.

**Comment by lornajane** [comment]
This was fixed in #3167
